### PR TITLE
Poll socket before attempting a speculative read or write

### PR DESCRIPTION
**Pull request by yourslab**
This addresses the problem from issue #1634. Specifically, the POSIX transport implementation is updated to match what is specified in the [transport_interface.h documentation for handling speculative reads](https://github.com/FreeRTOS/coreMQTT/blob/main/source/interface/transport_interface.h#L94-L172).

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: platform/posix/transport/src/openssl_posix.c [hunk 1]
```diff
@@ -729,55 +721,96 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
...
+        struct pollfd pollFds;
         pOpensslParams = pNetworkContext->pParams;
-        /* SSL read of data. */
-        bytesReceived = ( int32_t ) SSL_read( pOpensslParams->pSsl,
-                                              pBuffer,
-                                              ( int32_t ) bytesToRecv );
+
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLIN | POLLPRI | POLLHUP | POLLERR;
+        pollFds.revents = 0;
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:743]: A comment about the significance of these event flags being set can be helpful

File: platform/posix/transport/src/openssl_posix.c [hunk 2]
```diff
@@ -729,55 +721,96 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
...
+
+        if( ( bytesToRecv > 1 ) || ( SSL_pending( pOpensslParams->pSsl ) > 0 ) )
+        {
+            shouldRead = 1U;
+        }
+        else
+        {
+            /* Non-speculative read so check if there is data to read from the socket. */
+            pollStatus = poll( &pollFds, 1, 0 );
+        }
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:755]: Comments here to clearly mention that there is a non-blocking read for the case of reading more than one byte whereas  a timed wait when reading a single byte is the behavior represented by this if-else block

File: platform/posix/transport/src/openssl_posix.c [hunk 3]
```diff
@@ -729,55 +721,96 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
...
+        uint8_t shouldRead = 0U;
+        struct pollfd pollFds;
         pOpensslParams = pNetworkContext->pParams;
-        /* SSL read of data. */
-        bytesReceived = ( int32_t ) SSL_read( pOpensslParams->pSsl,
-                                              pBuffer,
-                                              ( int32_t ) bytesToRecv );
+
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLIN | POLLPRI | POLLHUP | POLLERR;
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:742]: According to the [Linux man page](https://man7.org/linux/man-pages/man2/poll.2.html) for `poll.h`:
"_The bits returned in revents can include any of those specified in events, or one of the values **POLLERR**, **POLLHUP**, or **POLLNVAL**.  (These three bits are meaningless in the events field, and will be set in the revents field whenever the corresponding condition is true.)_"

Thus. explicitly setting `POLLHUP | POLLERR` bits in the `events` member is not required..

File: platform/posix/transport/src/openssl_posix.c [hunk 4]
```diff
@@ -729,55 +721,96 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
...
+        if( pollStatus < 0 )
+        {
+            bytesReceived = -1;
+        }
+        else if( pollStatus == 0 )
+        {
+            /* No data available to be read from the socket. */
+            bytesReceived = 0;
+        }
+        else
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:766]: There should probably a check that the `pollFds.revent` output parameter has been set to `POLLIN` based on the following documentation for positive return value from `poll` in the [Linux man page](https://man7.org/linux/man-pages/man2/poll.2.html):
"_On success, poll() returns a nonnegative value which is the number of elements in the pollfds whose revents fields have been set to a nonzero value (indicating an event or an error).  A return value of zero indicates that the system call timed out before any file descriptors became read._"

Comment (aggarw13) [reply platform/posix/transport/src/openssl_posix.c:766]: Also, an appropriate log message can be added if the `POLLERR` flag is set in the output parameter of `revents`

File: platform/posix/transport/src/openssl_posix.c [hunk 5]
```diff
@@ -792,47 +825,70 @@ int32_t Openssl_Send( NetworkContext_t * pNetworkContext,
...
+
         pOpensslParams = pNetworkContext->pParams;
-        /* SSL write of data. */
-        bytesSent = ( int32_t ) SSL_write( pOpensslParams->pSsl,
-                                           pBuffer,
-                                           ( int32_t ) bytesToSend );
 
-        if( bytesSent <= 0 )
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLOUT | POLLPRI | POLLHUP | POLLERR;
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:841]: Same comments here as [mentioned earlier](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1665/files?file-filters%5B%5D=.c#r669855972) about not needing `POLLHUP`, `POLLERR` flags to be explicitly set here as these are always enabled

File: platform/posix/transport/src/openssl_posix.c (hunk 2 above)
Comment (yourslab) [reply platform/posix/transport/src/openssl_posix.c:755]: `SSL_pending` will only return a non-zero value after a call to `SSL_read`. It just refers to data that was not yet processed on the last read. It is still necessary to have a mechanism to check if the socket has data to be read.

File: platform/posix/transport/src/openssl_posix.c (hunk 4 above)
Comment (yourslab) [reply platform/posix/transport/src/openssl_posix.c:766]: We had a chat that the error in such cases would already be handled by `recv` for POSIX and `SSL_read` for OpenSSL, so it is not necessary to check these flags.

File: platform/posix/transport/src/openssl_posix.c [hunk 6]
```diff
@@ -792,47 +827,70 @@ int32_t Openssl_Send( NetworkContext_t * pNetworkContext,
...
         }
+        else if( pollStatus < 0 )
+        {
+            /* An error occurred while polling. */
+            LogError( ( "An error occurred while polling." ) );
+            bytesSent = -1;
+        }
+        else
+        {
+            /* Timed out waiting for data to be sent. */
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:891]: There was no timed wait for the socket to be available as the `timeout` parameter used is `0`. So, this comment needs to be updated
```suggestion
            /* Socket is not available for sending data. Set return code for retrying send. */
```

File: platform/posix/transport/src/openssl_posix.c (hunk 2 above)
Comment (aggarw13) [reply platform/posix/transport/src/openssl_posix.c:757]: Doesn't `SSL_pendign` data represent all the data buffered _after_ processing input records?

Comment (aggarw13) [reply platform/posix/transport/src/openssl_posix.c:757]: I see a problem if a partial SSL record is received from the network, but `poll` returns a positive return code to represent availability of data on the TCP socket, but `SSL_read` returns zero bytes because the entire SSL record data has not been received. In that case, this function shouldn't return negative return code back to the calling MQTT/HTTP library

File: platform/posix/transport/src/openssl_posix.c [hunk 7]
```diff
@@ -792,47 +827,70 @@ int32_t Openssl_Send( NetworkContext_t * pNetworkContext,
...
-                                           ( int32_t ) bytesToSend );
 
-        if( bytesSent <= 0 )
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLOUT;
+        pollFds.revents = 0;
+        /* Set the file descriptor for poll. */
+        pollFds.fd = pOpensslParams->socketDescriptor;
+
+        /* Check if socket descriptor is ready to be written to. */
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:848]: A comment here about the non-blocking call to `poll` will be valuable

File: platform/posix/transport/src/openssl_posix.c [hunk 8]
```diff
@@ -792,47 +827,70 @@ int32_t Openssl_Send( NetworkContext_t * pNetworkContext,
...
+                if( bytesSent == 0 )
+                {
+                    bytesSent = -1;
+                }
             }
         }
+        else if( pollStatus < 0 )
+        {
+            /* An error occurred while polling. */
+            LogError( ( "An error occurred while polling." ) );
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:886]:
```suggestion
            LogError( ( "Unable to send TLS data on network: An error occurred while checking availability of TCP socket, %d, to send data.",  pOpensslParams->socketDescriptor) );
```

File: platform/posix/transport/src/plaintext_posix.c [hunk 9]
```diff
@@ -115,79 +115,40 @@ int32_t Plaintext_Recv( NetworkContext_t * pNetworkContext,
...
-    /* Check if there is data to read from the socket. */
-    selectStatus = select( pPlaintextParams->socketDescriptor + 1,
-                           &readfds,
-                           NULL,
-                           NULL,
-                           &recvTimeout );
+    /* Speculative read for the start of a payload. */
+    if( bytesToRecv == 1U )
+    {
+        /* Check if there is data to read from the socket. */
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:139]:
```suggestion
        /* Check if there is data to read (without blocking) from the socket. */
```

File: platform/posix/transport/src/plaintext_posix.c [hunk 10]
```diff
@@ -115,79 +115,40 @@ int32_t Plaintext_Recv( NetworkContext_t * pNetworkContext,
...
+    /* Set the file descriptor for poll. */
+    pollFds.fd = pPlaintextParams->socketDescriptor;
 
-    /* Check if there is data to read from the socket. */
-    selectStatus = select( pPlaintextParams->socketDescriptor + 1,
-                           &readfds,
-                           NULL,
-                           NULL,
-                           &recvTimeout );
+    /* Speculative read for the start of a payload. */
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:136]:
```suggestion
    /* Speculative read for the start of a payload. */
    /* Note: This is done to avoid blocking behavior of recv() call when no data is available from the network to read. */
```

File: platform/posix/transport/src/plaintext_posix.c [hunk 11]
```diff
@@ -115,79 +115,40 @@ int32_t Plaintext_Recv( NetworkContext_t * pNetworkContext,
...
         bytesReceived = ( int32_t ) recv( pPlaintextParams->socketDescriptor,
                                           pBuffer,
                                           bytesToRecv,
                                           0 );
     }
-    else if( selectStatus < 0 )
+    else if( pollStatus < 0 )
     {
         /* An error occurred while polling. */
         bytesReceived = -1;
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:154]: The comment on line 158 is invalid about timeout from the `poll` call.

File: platform/posix/transport/src/plaintext_posix.c [hunk 12]
```diff
@@ -198,7 +159,7 @@ int32_t Plaintext_Recv( NetworkContext_t * pNetworkContext,
         bytesReceived = 0;
     }
 
-    if( ( selectStatus > 0 ) && ( bytesReceived == 0 ) )
+    if( ( pollStatus > 0 ) && ( bytesReceived == 0 ) )
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:162]: (Nit) A comment can be added to mention that a zero value return from `recv` represents closure of TCP connection by the peer. I know that this comment exists within this branch, but it doesn't explicitly mention about it being related to `recv` call

File: platform/posix/transport/src/plaintext_posix.c [hunk 13]
```diff
@@ -224,77 +185,34 @@ int32_t Plaintext_Send( NetworkContext_t * pNetworkContext,
...
-                           &sendTimeout );
-
-    if( selectStatus > 0 )
+    /* Initialize the file descriptor. */
+    pollFds.events = POLLOUT;
+    pollFds.revents = 0;
+    /* Set the file descriptor for poll. */
+    pollFds.fd = pPlaintextParams->socketDescriptor;
+
+    /* Check if there is data to read from the socket. */
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:204]:
```suggestion
    /* Check if socket is available to send data on the network. This is done to avoid blocking on send() call when the socket is ready to accept more data to send on the network (possibly due to a full TX buffer) . */
```

File: platform/posix/transport/src/plaintext_posix.c [hunk 14]
```diff
@@ -224,77 +185,34 @@ int32_t Plaintext_Send( NetworkContext_t * pNetworkContext,
...
         /* The socket is available for sending data. */
         bytesSent = ( int32_t ) send( pPlaintextParams->socketDescriptor,
                                       pBuffer,
                                       bytesToSend,
                                       0 );
     }
-    else if( selectStatus < 0 )
+    else if( pollStatus < 0 )
     {
         /* An error occurred while polling. */
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:217]: "Timed out" comment on line 222 is not valid anymore

File: platform/posix/transport/src/openssl_posix.c (hunk 2 above)
Comment (yourslab) [reply platform/posix/transport/src/openssl_posix.c:757]: We discussed that `poll` is necessary in this case because `SSL_pending` does not actually tell you if data is ready to be read from the socket, it will only return a value >0 if data remains to be read from the previous call to `SSL_read`.

File: platform/posix/transport/utest/openssl_utest.c [hunk 15]
```diff
@@ -848,10 +875,30 @@ void test_Openssl_Recv_Network_Error( void )
     TEST_ASSERT_TRUE( bytesReceived < 0 );
 
     /* Test that a non-retryable zero error code is converted to -1 by the API. */
+    SSL_pending_ExpectAnyArgsAndReturn( 1 );
     SSL_read_ExpectAnyArgsAndReturn( 0 );
     SSL_get_error_ExpectAnyArgsAndReturn( SSL_ERROR_ZERO_RETURN );
-    bytesReceived = Openssl_Recv( &networkContext, opensslBuffer, BYTES_TO_RECV );
+    bytesReceived = Openssl_Recv( &networkContext, opensslBuffer, 1U );
```
Comment (aggarw13) [inline platform/posix/transport/utest/openssl_utest.c:881]: Why was the number of bytes parameter changed?

File: platform/posix/transport/utest/openssl_utest.c [hunk 16]
```diff
@@ -821,10 +842,16 @@ void test_Openssl_Recv_All_Bytes_Received_Successfully( void )
...
     int32_t bytesReceived;
 
+    /* No pending data from last read. */
     opensslParams.pSsl = &ssl;
     SSL_read_ExpectAnyArgsAndReturn( BYTES_TO_RECV );
     bytesReceived = Openssl_Recv( &networkContext, opensslBuffer, BYTES_TO_RECV );
     TEST_ASSERT_EQUAL( BYTES_TO_RECV, bytesReceived );
+
+    /* Pending data from last read. */
+    SSL_read_ExpectAnyArgsAndReturn( BYTES_TO_RECV );
```
Comment (aggarw13) [inline platform/posix/transport/utest/openssl_utest.c:852]: This case seems to be an exact repetition of the case that already exists in this function..?

File: platform/posix/transport/utest/openssl_utest.c [hunk 17]
```diff
@@ -848,10 +875,30 @@ void test_Openssl_Recv_Network_Error( void )
...
+/**
+ * @brief Test that #Openssl_Recv returns 0 when `poll` returns 0.
+ */
+void test_Openssl_Recv_Poll_No_Events( void )
+{
+    int32_t bytesReceived;
+
+    SSL_pending_ExpectAnyArgsAndReturn( 0 );
+    poll_ExpectAnyArgsAndReturn( 0 );
+    bytesReceived = Openssl_Recv( &networkContext, opensslBuffer, 1U );
```
Comment (aggarw13) [inline platform/posix/transport/utest/openssl_utest.c:900]: Should this also test the case when more than one byte is requested but both `SSL_pending` and `poll` return zero?

Comment (yourslab) [reply platform/posix/transport/utest/openssl_utest.c:900]: Sure, doesn't hurt to have it

File: platform/posix/transport/utest/openssl_utest.c (hunk 16 above)
Comment (yourslab) [reply platform/posix/transport/utest/openssl_utest.c:852]: I had meant to change this but got coverage in another function. Changed that here.

File: platform/posix/transport/utest/openssl_utest.c (hunk 15 above)
Comment (yourslab) [reply platform/posix/transport/utest/openssl_utest.c:881]: I had forgotten to update the other test that would've attained the same coverage. I reverted this.

File: platform/posix/transport/utest/openssl_utest.c (hunk 17 above)
Comment (yourslab) [reply platform/posix/transport/utest/openssl_utest.c:900]: If more than one byte is requested, neither of those functions are called

File: platform/posix/transport/src/openssl_posix.c [hunk 18]
```diff
@@ -746,13 +746,17 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
...
         if( ( bytesToRecv > 1 ) || ( SSL_pending( pOpensslParams->pSsl ) > 0 ) )
         {
             shouldRead = 1U;
         }
         else
         {
-            /* Non-speculative read so check if there is data to read from the socket. */
+            /* Speculative read for the start of a payload.
+             * Note: This is done to avoid blocking when no
+             * data is available to be read from the socket. */
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:759]: Formatting of comment is not aligned to the code
```suggestion
           /* Speculative read for the start of a payload.
            * Note: This is done to avoid blocking when no
            * data is available to be read from the socket. */
```

File: platform/posix/transport/src/openssl_posix.c [hunk 19]
```diff
@@ -746,13 +746,17 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
         /* Set the file descriptor for poll. */
         pollFds.fd = pOpensslParams->socketDescriptor;
 
+        /* #SSL_pending returns a value > 0 if application data
+         * from the last processed TLS record remains to be read. */
         if( ( bytesToRecv > 1 ) || ( SSL_pending( pOpensslParams->pSsl ) > 0 ) )
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:751]: It could be fine to call `SSL_pending` for every call to this function, irrespective of the number of bytes requested. This can act as the first level of filter where the OpenSSL library is queried about pending data, followed by checking the `poll` function for data from the TCP socket

File: platform/posix/transport/src/plaintext_posix.c [hunk 20]
```diff
@@ -201,7 +205,10 @@ int32_t Plaintext_Send( NetworkContext_t * pNetworkContext,
     /* Set the file descriptor for poll. */
     pollFds.fd = pPlaintextParams->socketDescriptor;
 
-    /* Check if there is data to read from the socket. */
+    /* Check if data can be written to the socket.
+     * Note: This is done to avoid blocking on send() call when
+     * the socket is ready to accept more data for transmitting
+     * (possibly due to a full TX buffer). */
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:211]:
```suggestion
     * Note: This is done to avoid blocking on send() call when
     * the socket is not ready to accept more data for networl
     * transmission (possibly due to a full TX buffer). */
```

File: platform/posix/transport/utest/openssl_utest.c [hunk 21]
```diff
@@ -849,8 +849,9 @@ void test_Openssl_Recv_All_Bytes_Received_Successfully( void )
     TEST_ASSERT_EQUAL( BYTES_TO_RECV, bytesReceived );
 
     /* Pending data from last read. */
+    SSL_pending_ExpectAnyArgsAndReturn( BYTES_TO_RECV );
     SSL_read_ExpectAnyArgsAndReturn( BYTES_TO_RECV );
```
Comment (aggarw13) [inline platform/posix/transport/utest/openssl_utest.c:853]: Is there a reason for returning `BYTES_TO_RECV` from `SSL_read` when only a single byte is requested in the test?

Comment (yourslab) [reply platform/posix/transport/utest/openssl_utest.c:853]: I had forgotten to update it

File: platform/posix/transport/src/openssl_posix.c (hunk 18 above)
Comment (yourslab) [reply platform/posix/transport/src/openssl_posix.c:759]: That is already aligned. What you are suggesting won't pass formatting checks..

File: platform/posix/transport/src/plaintext_posix.c [hunk 22]
```diff
@@ -224,88 +189,48 @@ int32_t Plaintext_Send( NetworkContext_t * pNetworkContext,
...
+    /* Initialize the file descriptor. */
+    pollFds.events = POLLOUT;
+    pollFds.revents = 0;
+    /* Set the file descriptor for poll. */
+    pollFds.fd = pPlaintextParams->socketDescriptor;
+
+    /* Check if data can be written to the socket.
+     * Note: This is done to avoid blocking on send() when
+     * the socket is ready to accept more data for network transmission
+     * (possibly due to a full TX buffer). */
```
Comment (aggarw13) [inline platform/posix/transport/src/plaintext_posix.c:211]:
```suggestion
     * the socket is not ready to accept more data for network 
     * transmission (possibly due to a full TX buffer). */
```

File: platform/posix/transport/src/openssl_posix.c [hunk 23]
```diff
@@ -792,47 +838,72 @@ int32_t Openssl_Send( NetworkContext_t * pNetworkContext,
...
-                                           ( int32_t ) bytesToSend );
 
-        if( bytesSent <= 0 )
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLOUT;
+        pollFds.revents = 0;
+        /* Set the file descriptor for poll. */
+        pollFds.fd = pOpensslParams->socketDescriptor;
+
+        /* `poll` checks if the socket is ready to send data. */
```
Comment (aggarw13) [inline platform/posix/transport/src/openssl_posix.c:859]:
```suggestion
        /* `poll` checks if the socket is ready to send data. 
         * Note: This is done to avoid blocking on SSL_write()
         * when TCP socket is not ready to accept more data for
         * network transmission (possibly due to a full TX buffer). */
```

**Review by aggarw13 — DISMISSED** [review]

**Review by aggarw13 — APPROVED** [review]

File: platform/posix/transport/utest/openssl_utest.c [hunk 24]
```diff
@@ -785,10 +788,28 @@ void test_Openssl_Send_Network_Error( void )
     TEST_ASSERT_TRUE( bytesSent < 0 );
 
     /* Test that a non-retryable zero error code is converted to -1 by the API. */
+    poll_ExpectAnyArgsAndReturn( 1 );
     SSL_write_ExpectAnyArgsAndReturn( 0 );
     SSL_get_error_ExpectAnyArgsAndReturn( SSL_ERROR_ZERO_RETURN );
     bytesSent = Openssl_Send( &networkContext, opensslBuffer, BYTES_TO_SEND );
     TEST_ASSERT_EQUAL( -1, bytesSent );
+
+    /* Test that a poll error results in the function under test returning -1. */
```
Comment (paulbartell) [inline platform/posix/transport/utest/openssl_utest.c:797]: A test double would make this test less brittle.

**Review by paulbartell — APPROVED** [review]
